src/main/java/com/puppycrawl/tools/checkstyle/PackageObjectFactory.java

```diff
@@ -187,10 +188,11 @@ public Object createModule(String name) throws CheckstyleException {
             instance = createFromStandardCheckSet(name);
             // find the name in third party map
             if (instance == null) {
-                if (thirdPartyNameToFullModuleNames == null) {
-                    thirdPartyNameToFullModuleNames =
-                            generateThirdPartyNameToFullModuleName(moduleClassLoader);
-                }
+                final Optional<Map<String, Set<String>>> cache =
+                        Optional.ofNullable(thirdPartyNameToFullModuleNames);
```

Comment:
I am surprised that none of our static analysis caught this. IMO, this is abuse of `Optional`.  Collections are better to be empty than optional. Typically `Optional` is used for a single object. Can we do:
1. Make this field final 9`thirdPartyNameToFullModuleNames`)
2. Initialize to empty Collection in constructor
3. Populate the Collection in this method
?

Reply:
Let's do this in separate issue, we need to clean up pitest, project is close to completion.

Reply:
I would much rather we allow a suppression for pitest than allow Optional collections into the codebase. We have a mutable collection as a class field, this should be addressed anyway.

Reply:
> Can we do: ...

The mutation survives in this case. We either use Optional or use a wrapper method with exclusion in the pom.xml unless you have other ideas how to kill it.

Reply:
Optional is more elegant than another options

Reply:
IMO: best option is to leave this suppression with an explanation, this is why we have a suppression system.

Optional collection is a hack just like some wrapper method. We should not allow pitest to mangle production code like this; even a suppression in the pom is better than this. I don't see any reason to change code in cases where we have lazy loading, etc. for performance reasons. We are becoming too fanatic with our desire for empty suppression files.

The existing pitest suppression is the most targeted way to suppress this code; we should use it.

Reply:
The other option seems also just as easy, write a test to cover this situation.

The whole point of writing tests, besides coverage, is to also show expected behavior and functionality. It is expected we have lazy loading here to ensure we don't waste CPU cycles reloading the same data.

We have written tests using Mockito before and it provides the ability count how many times a method is called. We also use it in the releasenotes-builder tests.
https://stackoverflow.com/a/39452689/1016482

We should not be afraid to write these type of tests as this is the only way to prove this behavior is expected.

10 years down the road, no one will understand why we wrote such bad looking code if we merge this PR as it is. I am sorry, but I see this production code rewrite as just a method we have found to "outsmart" pitest, and have not actually solved the problem of showcasing lazy loading.

Reply:
Some ideas came to my mind.
Make method 
```
private initializeThirdPartyModuleMapIfRequired() {
if (thirdPartyNameToFullModuleNames == null) {
                    thirdPartyNameToFullModuleNames =
                            generateThirdPartyNameToFullModuleName(moduleClassLoader);
 }
}
```

And add it to pitest exclude in pom.xml

Reply:
If lazy initialization is a concern elsewhere, we could just make a generic util method that takes a supplier and use that.

Reply:
rethingking code in different way - https://github.com/checkstyle/checkstyle/pull/13744